### ISSUE TITLE
Incorrect error about assignment to borrowed value when using destructuring and conditional returns

### ISSUE THREAD

**Issue by athre0z** [rust-lang/rust#116499]
I expect this code to compile:

```rust
struct Foo {
    blah: i32
}

fn some_fn(x: &mut Foo) -> &i32 {
    if x.blah == 2 {
        let Foo { blah } = x;
        
        if 333 == 444 {
            return blah;
        }
    }
    
    x.blah = 333;
    &x.blah
}
```

But I'm getting an E0506:

```
error[E0506]: cannot assign to `x.blah` because it is borrowed
  --> src/main.rs:14:5
   |
5  | fn some_fn(x: &mut Foo) -> &i32 {
   |               - let's call the lifetime of this reference `'1`
6  |     if x.blah == 2 {
7  |         let Foo { blah } = x;
   |                   ---- `x.blah` is borrowed here
...
10 |             return blah;
   |                    ---- returning this value requires that `x.blah` is borrowed for `'1`
...
14 |     x.blah = 333;
   |     ^^^^^^^^^^^^ `x.blah` is assigned to here but it was already borrowed

For more information about this error, try `rustc --explain E0506`.
error: could not compile `playground` (bin "playground") due to previous error
```

If the second level of `if` is commented out (lifting the return out into the first level of `if), it compiles fine. 

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b036f408b79716b02923b9ee8bb67f00)

rustc versions tried:

- local rustc 1.75.0-nightly (e0d7ed1f4 2023-10-01)
- stable from playground (1.73)
- nightly from playground (2023-10-05)
